From 9b9964449c519ad68d02e9c98d151d393975425c Mon Sep 17 00:00:00 2001
From: Bastian Bittorf <bittorf@bluebottle.com>
Date: Tue, 24 Nov 2015 16:23:25 +0100
Subject: [PATCH] busybox: Makefile: bump to git head from 2015-nov-6


Signed-off-by: Bastian Bittorf <bittorf@bluebottle.com>
---
 package/utils/busybox/Makefile | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/package/utils/busybox/Makefile b/package/utils/busybox/Makefile
index 5ca4363..b0da848 100644
--- a/package/utils/busybox/Makefile
+++ b/package/utils/busybox/Makefile
@@ -8,14 +8,16 @@
 include $(TOPDIR)/rules.mk
 
 PKG_NAME:=busybox
-PKG_VERSION:=1.24.1
+PKG_SOURCE_VERSION:=196e400441652946b9c7ad7bc2d78c73885f2359
+PKG_VERSION:=2015-11-06-$(PKG_SOURCE_VERSION)
 PKG_RELEASE:=1
 PKG_FLAGS:=essential
 
-PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.bz2
-PKG_SOURCE_URL:=http://www.busybox.net/downloads \
-		http://distfiles.gentoo.org/distfiles/
-PKG_MD5SUM:=be98a40cadf84ce2d6b05fa41a275c6a
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.xz
+PKG_SOURCE_SUBDIR:=$(PKG_NAME)-$(PKG_VERSION)
+PKG_SOURCE_URL:=git://git.busybox.net/busybox
+PKG_MIRROR_MD5SUM:=ca4d289ad9380471cae376fc7dd3660a
 
 PKG_BUILD_DEPENDS:=BUSYBOX_USE_LIBRPC:librpc BUSYBOX_CONFIG_PAM:libpam
 PKG_BUILD_PARALLEL:=1
-- 
1.8.1.2

