From 7677fb106c2868b1cff7c5428010274bbd556f71 Mon Sep 17 00:00:00 2001
From: Bastian Bittorf <bittorf@bluebottle.com>
Date: Wed, 19 Aug 2015 15:20:32 +0200
Subject: [PATCH] dropbear: allow cipher NONE


Signed-off-by: Bastian Bittorf <bittorf@bluebottle.com>
---
 package/network/services/dropbear/Makefile |    3 +++
 1 file changed, 3 insertions(+)

diff --git a/package/network/services/dropbear/Makefile b/package/network/services/dropbear/Makefile
index f793e9b..91f55ea 100644
--- a/package/network/services/dropbear/Makefile
+++ b/package/network/services/dropbear/Makefile
@@ -90,6 +90,9 @@ define Build/Configure
 	       >$(PKG_BUILD_DIR)/options.h.new && \
 	  mv $(PKG_BUILD_DIR)/options.h.new $(PKG_BUILD_DIR)/options.h || exit 1; \
 	done
+
+	grep -q ^'#define DROPBEAR_NONE_CIPHER' $(PKG_BUILD_DIR)/options.h || \
+		$(SED) 's|.*DROPBEAR_NONE_CIPHER.*|#define DROPBEAR_NONE_CIPHER|' $(PKG_BUILD_DIR)/options.h
 endef
 
 define Build/Compile
-- 
1.7.10.4

