From 0e6ec73cdf3479e4a068401f86e7bcc8c948cbed Mon Sep 17 00:00:00 2001
From: Bastian Bittorf <bittorf@bluebottle.com>
Date: Mon, 29 Feb 2016 10:38:37 +0100
Subject: [PATCH] more debug

Signed-off-by: Bastian Bittorf <bittorf@bluebottle.com>
---
 libfstools/overlay.c | 23 ++++++++++++++++++++---
 1 file changed, 20 insertions(+), 3 deletions(-)

diff --git a/libfstools/overlay.c b/libfstools/overlay.c
index cdac23e..620db4a 100644
--- a/libfstools/overlay.c
+++ b/libfstools/overlay.c
@@ -153,6 +153,7 @@ switch2jffs(struct volume *v)
 		return -1;
 	}
 
+	ULOG_ERR("return fopivot /overlay -> /rom\n");
 	return fopivot("/overlay", "/rom");
 }
 
@@ -228,12 +229,24 @@ jffs2_switch(struct volume *v)
 
 	case FS_JFFS2:
 		ret = overlay_mount(v, "jffs2");
-		if (ret)
+		if (ret) {
+			ULOG_ERR("FS_JFFS2: overlay_mount OK\n");
 			break;
-		if (mount_move("/tmp", "", "/overlay") || fopivot("/overlay", "/rom")) {
-			ULOG_ERR("switching to jffs2 failed\n");
+		}
+		ULOG_ERR("FS_JFFS2: overlay_mount failed\n");
+
+		if (mount_move("/tmp", "", "/overlay")) {
+			ULOG_ERR("FS_JFFS2: switching to jffs2 OK: mount_move: /tmp -> '' -> /overlay\n");
 			ret = -1;
 		}
+		ULOG_ERR("FS_JFFS2: switching to jffs2 failed: mount_move: /tmp -> '' -> /overlay\n");
+
+		if (fopivot("/overlay", "/rom")) {
+			ULOG_ERR("FS_JFFS2: switching to jffs2 OK: fopivot: '/overlay' -> '/rom'\n");
+			ret = -1;
+		}
+		ULOG_ERR("FS_JFFS2: switching to jffs2 failed: fopivot: '/overlay' -> '/rom'\n");
+
 		break;
 
 	case FS_UBIFS:
@@ -247,11 +260,15 @@ jffs2_switch(struct volume *v)
 		break;
 	}
 
+	ULOG_ERR("isok?");
 	if (ret)
 		return ret;
 
+	ULOG_ERR("sync start");
 	sync();
+	ULOG_ERR("sync ready, fs_state_set /overlay -> FS_STATE_READY");
 	fs_state_set("/overlay", FS_STATE_READY);
+	ULOG_ERR("fs_state_set ok");
 	return 0;
 }
 
-- 
2.1.4

