From 346191dfa453ee0e4d48945cb65e91b35a90f2b6 Mon Sep 17 00:00:00 2001
From: Bastian Bittorf <bittorf@bluebottle.com>
Date: Tue, 24 Nov 2015 16:23:25 +0100
Subject: [PATCH] busybox: Makefile: bump to git head from 2015-nov-24


Signed-off-by: Bastian Bittorf <bittorf@bluebottle.com>
---
 package/utils/busybox/Makefile | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/package/utils/busybox/Makefile b/package/utils/busybox/Makefile
index 5ca4363..fb17cf2 100644
--- a/package/utils/busybox/Makefile
+++ b/package/utils/busybox/Makefile
@@ -8,14 +8,13 @@
 include $(TOPDIR)/rules.mk
 
 PKG_NAME:=busybox
-PKG_VERSION:=1.24.1
+PKG_VERSION:=20151124
 PKG_RELEASE:=1
 PKG_FLAGS:=essential
 
 PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.bz2
-PKG_SOURCE_URL:=http://www.busybox.net/downloads \
-		http://distfiles.gentoo.org/distfiles/
-PKG_MD5SUM:=be98a40cadf84ce2d6b05fa41a275c6a
+PKG_SOURCE_URL:=http://busybox.net/downloads/snapshots/
+PKG_MD5SUM:=ee945b51b013decd2ab7b66916a911de
 
 PKG_BUILD_DEPENDS:=BUSYBOX_USE_LIBRPC:librpc BUSYBOX_CONFIG_PAM:libpam
 PKG_BUILD_PARALLEL:=1
-- 
1.8.1.2

