From 283c46fb06407e45e436e86d55c1f58471e9a3f7 Mon Sep 17 00:00:00 2001
From: Bastian Bittorf <bastian@bastian-laptop.(none)>
Date: Sat, 29 Jan 2011 13:46:19 +0100
Subject: [PATCH] recognize Dell Truemobile 2300 v1 / TM2300 with missing nvram var ModelId

---
 package/broadcom-diag/src/diag.c |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

diff --git a/package/broadcom-diag/src/diag.c b/package/broadcom-diag/src/diag.c
index 55dfa75..4c4b432 100644
--- a/package/broadcom-diag/src/diag.c
+++ b/package/broadcom-diag/src/diag.c
@@ -959,7 +959,7 @@ static struct platform_t __init *platform_detect(void)
 		if (!strcmp(buf, "WR850G"))
 			return &platforms[WR850GV2V3];
 		if (!strcmp(buf, "WX-5565") && !strcmp(getvar("boardtype"),"bcm94710ap"))
-			return &platforms[TM2300]; /* Dell TrueMobile 2300 */
+			return &platforms[TM2300]; /* Dell TrueMobile 2300 v1 */
 		if (startswith(buf,"WE800G")) /* WE800G* */
 			return &platforms[WE800G];
 	}
@@ -1077,6 +1077,9 @@ static struct platform_t __init *platform_detect(void)
 			if (simple_strtoul(boardnum, NULL, 0) == 2)
 				return &platforms[WAP54GV1];
 		}
+		if (!strcmp(boardnum, "44") && !strcmp(boardtype,"bcm94710ap"))
+			return &platforms[TM2300]; /* Dell TrueMobile 2300 v1 (ModelId unset) */
+
 		/* MN-700 has also hardware_version 'WL500-...', so use boardnum */
 		if (startswith(getvar("hardware_version"), "WL500-")) {
 			if (!strcmp(getvar("boardnum"), "mn700"))
-- 
1.6.3.3

