--- package/feeds/luci/luci-addons/Makefile.orig	2014-10-04 15:37:09.026642377 +0200
+++ package/feeds/luci/luci-addons/Makefile	2014-10-04 15:37:18.583237754 +0200
@@ -95,7 +95,7 @@
 endef
 
 
-$(eval $(call module,freifunk,LuCI Freifunk module,+luci-mod-admin-full +luci-lib-json +PACKAGE_luci-mod-freifunk:freifunk-firewall +freifunk-common))
+$(eval $(call module,freifunk,LuCI Freifunk module,+luci-mod-admin-full +luci-lib-json +PACKAGE_luci-mod-freifunk: +freifunk-common))

