--- include/target.mk	2014-10-06 20:31:26.281611390 +0200
+++ include/target.mk.orig	2014-10-06 20:30:57.365187737 +0200
@@ -14,7 +14,7 @@
 # Default packages - the really basic set
 DEFAULT_PACKAGES:=base-files libc libgcc busybox dropbear mtd uci opkg netifd fstools
 # For router targets
-DEFAULT_PACKAGES.router:=dnsmasq iptables ip6tables kmod-ipt-nathelper odhcpd odhcp6c
+DEFAULT_PACKAGES.router:=dnsmasq iptables ip6tables ppp ppp-mod-pppoe kmod-ipt-nathelper firewall odhcpd odhcp6c
 DEFAULT_PACKAGES.bootloader:=
 
 ifneq ($(DUMP),)
