Index: compat-wireless-2014-11-04/net/mac80211/Kconfig
===================================================================
--- compat-wireless-2014-11-04.orig/net/mac80211/Kconfig	2015-02-13 01:28:16.561317337 +0100
+++ compat-wireless-2014-11-04/net/mac80211/Kconfig	2015-02-13 01:28:16.557317256 +0100
@@ -40,12 +40,36 @@
 	---help---
 	  This option enables VHT in the 'minstrel_ht' TX rate control algorithm
 
+config MAC80211_RC_RHAPSODY_BLUES
+	bool "Rhapsody-Blues 802.11b/g/a support" if EXPERT
+	select MAC80211_HAS_RC
+	default y
+	---help---
+	  This option enables the 'Rhapsody-Blues' joint TX rate & power control
+	  algorithm for IEEE 802.11b/g/a.
+
+config MAC80211_RC_RHAPSODY_BLUES_HT
+	bool "Rhapsody-Blues 802.11n support" if EXPERT
+	depends on MAC80211_RC_RHAPSODY_BLUES
+	default y
+	---help---
+	  This option enables the 'Rhapsody-Blues-HT' joint TX rate & power
+	  control algorithm for IEEE 802.11n.
+
+config MAC80211_RC_RHAPSODY_BLUES_VHT
+	bool "Rhapsody-Blues 802.11ac support" if EXPERT
+	depends on MAC80211_RC_RHAPSODY_BLUES_HT
+	default n
+	---help---
+	  This option enables the 'Rhapsody-Blues-VHT' joint TX rate & power
+	  control algorithm for IEEE 802.11ac.
+
 choice
-	prompt "Default rate control algorithm"
+	prompt "Default rate/power control algorithm"
 	depends on MAC80211_HAS_RC
 	default MAC80211_RC_DEFAULT_MINSTREL
 	---help---
-	  This option selects the default rate control algorithm
+	  This option selects the default rate/power control algorithm
 	  mac80211 will use. Note that this default can still be
 	  overridden through the ieee80211_default_rc_algo module
 	  parameter if different algorithms are available.
@@ -56,6 +80,12 @@
 	---help---
 	  Select Minstrel as the default rate control algorithm.
 
+config MAC80211_RC_DEFAULT_RHAPSODY_BLUES
+	bool "Rhapsody-Blues"
+	depends on MAC80211_RC_RHAPSODY_BLUES
+	---help---
+	  Select Rhapsody-Blues as the default joint rate & power control
+	  algorithm.
 
 endchoice
 
@@ -63,6 +93,8 @@
 	string
 	default "minstrel_ht" if MAC80211_RC_DEFAULT_MINSTREL && MAC80211_RC_MINSTREL_HT
 	default "minstrel" if MAC80211_RC_DEFAULT_MINSTREL
+	default "rhapsody_ht" if MAC80211_RC_DEFAULT_RHAPSODY_BLUES && MAC80211_RC_RHAPSODY_BLUES_HT
+	default "rhapsody" if MAC80211_RC_DEFAULT_RHAPSODY_BLUES
 	default ""
 
 endif
Index: compat-wireless-2014-11-04/net/mac80211/Makefile
===================================================================
--- compat-wireless-2014-11-04.orig/net/mac80211/Makefile	2015-02-13 01:28:16.561317337 +0100
+++ compat-wireless-2014-11-04/net/mac80211/Makefile	2015-02-13 01:28:16.557317256 +0100
@@ -58,4 +58,13 @@
 mac80211-$(CPTCFG_MAC80211_RC_MINSTREL) += $(rc80211_minstrel-y)
 mac80211-$(CPTCFG_MAC80211_RC_MINSTREL_HT) += $(rc80211_minstrel_ht-y)
 
+rc80211_rhapsody-y := rc80211_rhapsody.o
+rc80211_rhapsody-$(CPTCFG_MAC80211_DEBUGFS) += rc80211_rhapsody_debugfs.o
+
+rc80211_rhapsody_ht-y := rc80211_rhapsody_ht.o
+rc80211_rhapsody_ht-$(CPTCFG_MAC80211_DEBUGFS) += rc80211_rhapsody_ht_debugfs.o
+
+mac80211-$(CPTCFG_MAC80211_RC_RHAPSODY_BLUES) += $(rc80211_rhapsody-y)
+mac80211-$(CPTCFG_MAC80211_RC_RHAPSODY_BLUES_HT) += $(rc80211_rhapsody_ht-y)
+
 ccflags-y += -D__CHECK_ENDIAN__ -DDEBUG
Index: compat-wireless-2014-11-04/net/mac80211/rate.h
===================================================================
--- compat-wireless-2014-11-04.orig/net/mac80211/rate.h	2015-02-13 01:28:16.561317337 +0100
+++ compat-wireless-2014-11-04/net/mac80211/rate.h	2015-02-13 01:28:16.557317256 +0100
@@ -190,5 +190,30 @@
 }
 #endif
 
+#ifdef CPTCFG_MAC80211_RC_RHAPSODY_BLUES
+int rc80211_rhapsody_init(void);
+void rc80211_rhapsody_exit(void);
+#else
+static inline int rc80211_rhapsody_init(void)
+{
+	return 0;
+}
+static inline void rc80211_rhapsody_exit(void)
+{
+}
+#endif
+
+#ifdef CPTCFG_MAC80211_RC_RHAPSODY_BLUES_HT
+int rc80211_rhapsody_ht_init(void);
+void rc80211_rhapsody_ht_exit(void);
+#else
+static inline int rc80211_rhapsody_ht_init(void)
+{
+	return 0;
+}
+static inline void rc80211_rhapsody_ht_exit(void)
+{
+}
+#endif
 
 #endif /* IEEE80211_RATE_H */
Index: compat-wireless-2014-11-04/net/mac80211/main.c
===================================================================
--- compat-wireless-2014-11-04.orig/net/mac80211/main.c	2015-02-13 01:28:16.561317337 +0100
+++ compat-wireless-2014-11-04/net/mac80211/main.c	2015-02-13 01:28:16.557317256 +0100
@@ -1196,7 +1196,15 @@
 	BUILD_BUG_ON(sizeof(struct ieee80211_tx_info) > sizeof(skb->cb));
 	BUILD_BUG_ON(offsetof(struct ieee80211_tx_info, driver_data) +
 		     IEEE80211_TX_INFO_DRIVER_DATA_SIZE > sizeof(skb->cb));
+#ifdef CPTCFG_MAC80211_RC_RHAPSODY_BLUES
+	ret = rc80211_rhapsody_init();
+	if (ret)
+		return ret;
 
+	ret = rc80211_rhapsody_ht_init();
+	if (ret)
+		goto err_rhapsody;
+#endif
 	ret = rc80211_minstrel_init();
 	if (ret)
 		return ret;
@@ -1210,7 +1218,13 @@
 		goto err_netdev;
 
 	return 0;
+
  err_netdev:
+#ifdef CPTCFG_MAC80211_RC_RHAPSODY_BLUES
+	rc80211_rhapsody_ht_exit();
+ err_rhapsody:
+	rc80211_rhapsody_exit();
+#endif
 	rc80211_minstrel_ht_exit();
  err_minstrel:
 	rc80211_minstrel_exit();
@@ -1220,6 +1234,10 @@
 
 static void __exit ieee80211_exit(void)
 {
+#ifdef CPTCFG_MAC80211_RC_RHAPSODY_BLUES
+	rc80211_rhapsody_ht_exit();
+	rc80211_rhapsody_exit();
+#endif
 	rc80211_minstrel_ht_exit();
 	rc80211_minstrel_exit();
 
@@ -1236,3 +1254,4 @@
 
 MODULE_DESCRIPTION("IEEE 802.11 subsystem");
 MODULE_LICENSE("GPL");
+
Index: compat-wireless-2014-11-04/.local-symbols
===================================================================
--- compat-wireless-2014-11-04.orig/.local-symbols	2015-02-13 01:28:15.941304775 +0100
+++ compat-wireless-2014-11-04/.local-symbols	2015-02-13 01:29:30.686819136 +0100
@@ -41,6 +41,10 @@
 MAC80211_RC_MINSTREL_HT=
 MAC80211_RC_MINSTREL_VHT=
 MAC80211_RC_DEFAULT_MINSTREL=
+MAC80211_RC_RHAPSODY_BLUES=
+MAC80211_RC_RHAPSODY_BLUES_HT=
+MAC80211_RC_RHAPSODY_BLUES_VHT=
+MAC80211_RC_DEFAULT_RHAPSODY_BLUES=
 MAC80211_RC_DEFAULT=
 MAC80211_MESH=
 MAC80211_LEDS=
