Index: compat-wireless-2014-11-04/net/mac80211/rc80211_rhapsody.c
===================================================================
--- compat-wireless-2014-11-04.orig/net/mac80211/rc80211_rhapsody.c	2015-02-13 20:19:26.343758558 +0100
+++ compat-wireless-2014-11-04/net/mac80211/rc80211_rhapsody.c	2015-02-13 20:19:26.339758477 +0100
@@ -147,8 +147,8 @@
 {
 	int j = MAX_THR_RATES;
 
-	while (j > 0 && (rhapsody_get_tp_avg(&ri->r[i]) >
-				rhapsody_get_tp_avg(&ri->r[tp_list[j - 1]])))
+	while (j > 0 && (rhapsody_get_tp_5_percent_quantile(&ri->r[i]) >
+			 rhapsody_get_tp_5_percent_quantile(&ri->r[tp_list[j - 1]])))
 		j--;
 	if (j < MAX_THR_RATES - 1)
 		memmove(&tp_list[j + 1], &tp_list[j], MAX_THR_RATES - (j + 1));
@@ -156,6 +156,34 @@
 		tp_list[j] = i;
 }
 
+/* find & return topmost probability rate */
+static inline int
+rhapsody_get_best_prob_rate(struct rhapsody_sta_info *ri, int i)
+{
+	struct rhapsody_rate *rr = &ri->r[i];
+	int cur_tp, prob_tp;
+	u8 tmp_prob_rate = 0;
+
+	/* To determine the most robust rate (max_prob_rate) used at
+	 * 3rd mmr stage we distinct between two cases:
+	 * (1) if any success probabilitiy >= 95%, out of those rates
+	 * choose the maximum throughput rate as max_prob_rate
+	 * (2) if all success probabilities < 95%, the rate with
+	 * highest success probability is chosen as max_prob_rate */
+	if (rhapsody_get_prob_5_percent_quantile(rr) >= RHAPSODY_FRAC(95,100)) {
+		cur_tp = rhapsody_get_tp_5_percent_quantile(rr);
+		prob_tp = rhapsody_get_tp_5_percent_quantile(&ri->r[tmp_prob_rate]);
+		if (cur_tp >= prob_tp)
+			tmp_prob_rate = i;
+	} else {
+		if (rhapsody_get_prob_5_percent_quantile(rr) >=
+		    rhapsody_get_prob_5_percent_quantile(&ri->r[tmp_prob_rate]))
+			tmp_prob_rate = i;
+	}
+
+	return tmp_prob_rate;
+}
+
 static void
 rhapsody_set_rate(struct rhapsody_sta_info *ri, struct ieee80211_sta_rates *ratetbl,
 		  int offset, int idx)
@@ -240,7 +268,7 @@
 {
 	u8 tmp_tp_rate[MAX_THR_RATES];
 	u8 tmp_prob_rate = 0;
-	int i, tmp_cur_tp, tmp_prob_tp;
+	int i;
 
 	for (i = 0; i < MAX_THR_RATES; i++)
 	    tmp_tp_rate[i] = 0;
@@ -267,23 +295,9 @@
 		if (!rr->adjusted_retry_count)
 			rr->adjusted_retry_count = 2;
 
+		/* find fastest rate set and most robust rate */
 		rhapsody_sort_best_tp_rates(ri, i, tmp_tp_rate);
-
-		/* To determine the most robust rate (max_prob_rate) used at
-		 * 3rd mmr stage we distinct between two cases:
-		 * (1) if any success probabilitiy >= 95%, out of those rates
-		 * choose the maximum throughput rate as max_prob_rate
-		 * (2) if all success probabilities < 95%, the rate with
-		 * highest success probability is chosen as max_prob_rate */
-		if (rrs->prob_ewma >= RHAPSODY_FRAC(95, 100)) {
-			tmp_cur_tp = rhapsody_get_tp_avg(rr);
-			tmp_prob_tp = rhapsody_get_tp_avg(&ri->r[tmp_prob_rate]);
-			if (tmp_cur_tp >= tmp_prob_tp)
-				tmp_prob_rate = i;
-		} else {
-			if (rrs->prob_ewma >= ri->r[tmp_prob_rate].stats.prob_ewma)
-				tmp_prob_rate = i;
-		}
+		tmp_prob_rate = rhapsody_get_best_prob_rate(ri, i);
 	}
 
 	/* Assign the new rate set */
Index: compat-wireless-2014-11-04/net/mac80211/rc80211_rhapsody_ht.c
===================================================================
--- compat-wireless-2014-11-04.orig/net/mac80211/rc80211_rhapsody_ht.c	2015-02-13 20:19:26.343758558 +0100
+++ compat-wireless-2014-11-04/net/mac80211/rc80211_rhapsody_ht.c	2015-02-13 20:19:26.339758477 +0100
@@ -455,14 +455,20 @@
 
 	cur_group = index / MCS_GROUP_RATES;
 	cur_idx = index  % MCS_GROUP_RATES;
-	cur_tp_avg = rhapsody_ht_get_tp_avg(ri, cur_group, cur_idx);
-	cur_prob = ri->groups[cur_group].rates[cur_idx].prob_ewma;
+	cur_tp_avg = rhapsody_ht_get_tp_5_percent_quantile(ri, cur_group,
+							   cur_idx);
+	cur_prob = rhapsody_ht_get_prob_5_percent_quantile(ri, cur_group,
+							   cur_idx);
 
 	do {
 		tmp_group = tp_list[j - 1] / MCS_GROUP_RATES;
 		tmp_idx = tp_list[j - 1] % MCS_GROUP_RATES;
-		tmp_tp_avg = rhapsody_ht_get_tp_avg(ri, tmp_group, tmp_idx);
-		tmp_prob = ri->groups[tmp_group].rates[tmp_idx].prob_ewma;
+		tmp_tp_avg = rhapsody_ht_get_tp_5_percent_quantile(ri,
+								   tmp_group,
+								   tmp_idx);
+		tmp_prob = rhapsody_ht_get_prob_5_percent_quantile(ri,
+								   tmp_group,
+								   cur_idx);
 		if (cur_tp_avg < tmp_tp_avg ||
 		    (cur_tp_avg == tmp_tp_avg && cur_prob <= tmp_prob))
 			break;
@@ -484,21 +490,23 @@
 rhapsody_ht_set_best_prob_rate(struct rhapsody_ht_sta *ri, u16 index)
 {
 	struct rhapsody_mcs_group_data *rg;
-	struct rhapsody_rate_stats *rrs;
-	int tmp_group, tmp_idx, tmp_tp_avg, tmp_prob;
-	int max_tp_group, cur_tp_avg, cur_group, cur_idx;
+	int tmp_group, tmp_idx, tmp_tp, tmp_prob;
+	int max_tp_group, cur_tp, cur_group, cur_idx, cur_prob;
 	int max_group_prob_rate_group, max_group_prob_rate_idx;
-	int max_group_prob_rate_tp_avg;
+	int max_group_prob_rate_tp, max_group_prob_rate_prob;
 
+	/* candidate rate that_is compared to the current_max_prob one */
 	cur_group = index / MCS_GROUP_RATES;
 	cur_idx = index % MCS_GROUP_RATES;
 	rg = &ri->groups[index / MCS_GROUP_RATES];
-	rrs = &rg->rates[index % MCS_GROUP_RATES];
-
+	cur_prob = rhapsody_ht_get_prob_5_percent_quantile(ri, cur_group,
+							   cur_idx);
+	/* current max_prob_rate */
 	tmp_group = ri->max_prob_rate / MCS_GROUP_RATES;
 	tmp_idx = ri->max_prob_rate % MCS_GROUP_RATES;
-	tmp_tp_avg = rhapsody_ht_get_tp_avg(ri, tmp_group, tmp_idx);
-	tmp_prob = ri->groups[tmp_group].rates[tmp_idx].prob_ewma;
+	tmp_tp = rhapsody_ht_get_tp_5_percent_quantile(ri, tmp_group, tmp_idx);
+	tmp_prob = rhapsody_ht_get_prob_5_percent_quantile(ri, tmp_group,
+							   tmp_idx);
 
 	/* if max_tp_rate[0] is from MCS_GROUP max_prob_rate get selected from
 	 * MCS_GROUP as well as CCK_GROUP rates do not allow aggregation */
@@ -507,24 +515,34 @@
 	    (max_tp_group != RHAPSODY_CCK_GROUP))
 		return;
 
-	if (rrs->prob_ewma > RHAPSODY_FRAC(75, 100)) {
-		cur_tp_avg = rhapsody_ht_get_tp_avg(ri, cur_group, cur_idx);
-		if (cur_tp_avg > tmp_tp_avg)
+	/* if success prob. > 75% sort best_prob_rate by throughput */
+	if (cur_prob > 750) {
+		cur_tp = rhapsody_ht_get_tp_5_percent_quantile(ri, cur_group,
+							       cur_idx);
+		if (cur_tp > tmp_tp)
 			ri->max_prob_rate = index;
 
 		max_group_prob_rate_group = rg->max_group_prob_rate /
 								MCS_GROUP_RATES;
 		max_group_prob_rate_idx = rg->max_group_prob_rate %
 								MCS_GROUP_RATES;
-		max_group_prob_rate_tp_avg = rhapsody_ht_get_tp_avg(ri,
-						max_group_prob_rate_group,
-						max_group_prob_rate_idx);
-		if (cur_tp_avg > max_group_prob_rate_tp_avg)
+		max_group_prob_rate_tp = rhapsody_ht_get_tp_5_percent_quantile(ri,
+							max_group_prob_rate_group,
+							max_group_prob_rate_idx);
+		if (cur_tp > max_group_prob_rate_tp)
 			rg->max_group_prob_rate = index;
 	} else {
-		if (rrs->prob_ewma > tmp_prob)
+		if (cur_prob > tmp_prob)
 			ri->max_prob_rate = index;
-		if (rrs->prob_ewma > rg->rates[rg->max_group_prob_rate].prob_ewma)
+
+		max_group_prob_rate_group = rg->max_group_prob_rate /
+								MCS_GROUP_RATES;
+		max_group_prob_rate_idx = rg->max_group_prob_rate %
+								MCS_GROUP_RATES;
+		max_group_prob_rate_prob = rhapsody_ht_get_prob_5_percent_quantile(ri,
+							max_group_prob_rate_group,
+							max_group_prob_rate_idx);
+		if (cur_prob > max_group_prob_rate_prob)
 			rg->max_group_prob_rate = index;
 	}
 }
@@ -546,11 +564,13 @@
 
 	tmp_group = tmp_cck_tp_rate[0] / MCS_GROUP_RATES;
 	tmp_idx = tmp_cck_tp_rate[0] % MCS_GROUP_RATES;
-	tmp_cck_tp = rhapsody_ht_get_tp_avg(ri, tmp_group, tmp_idx);
+	tmp_cck_tp = rhapsody_ht_get_tp_5_percent_quantile(ri, tmp_group,
+							   tmp_idx);
 
 	tmp_group = tmp_mcs_tp_rate[0] / MCS_GROUP_RATES;
 	tmp_idx = tmp_mcs_tp_rate[0] % MCS_GROUP_RATES;
-	tmp_mcs_tp = rhapsody_ht_get_tp_avg(ri, tmp_group, tmp_idx);
+	tmp_mcs_tp = rhapsody_ht_get_tp_5_percent_quantile(ri, tmp_group,
+							   tmp_idx);
 
 	if (tmp_cck_tp > tmp_mcs_tp) {
 		for(i = 0; i < MAX_THR_RATES; i++) {
@@ -558,7 +578,6 @@
 						       tmp_mcs_tp_rate);
 		}
 	}
-
 }
 
 /*
@@ -569,7 +588,7 @@
 rhapsody_ht_prob_rate_reduce_streams(struct rhapsody_ht_sta *ri)
 {
 	struct rhapsody_mcs_group_data *rg;
-	int tmp_max_streams, group, tmp_idx;
+	int tmp_max_streams, group, cur_idx, cur_tp;
 	int tmp_tp = 0;
 
 	tmp_max_streams = rhapsody_mcs_groups[ri->max_tp_rate[0] /
@@ -579,13 +598,13 @@
 		if (!rg->supported || group == RHAPSODY_CCK_GROUP)
 			continue;
 
-		tmp_idx = rg->max_group_prob_rate % MCS_GROUP_RATES;
+		cur_idx = rg->max_group_prob_rate % MCS_GROUP_RATES;
+		cur_tp = rhapsody_ht_get_tp_5_percent_quantile(ri, group, cur_idx);
 
-		if (tmp_tp < rhapsody_ht_get_tp_avg(ri, group, tmp_idx) &&
+		if (tmp_tp < cur_tp &&
 		   (rhapsody_mcs_groups[group].streams < tmp_max_streams)) {
 				ri->max_prob_rate = rg->max_group_prob_rate;
-				tmp_tp = rhapsody_ht_get_tp_avg(ri, group,
-								tmp_idx);
+				tmp_tp = cur_tp;
 		}
 	}
 }
