Index: compat-wireless-2014-11-04/net/mac80211/rc80211_rhapsody_ht.c
===================================================================
--- compat-wireless-2014-11-04.orig/net/mac80211/rc80211_rhapsody_ht.c	2015-02-13 22:03:44.514192215 +0100
+++ compat-wireless-2014-11-04/net/mac80211/rc80211_rhapsody_ht.c	2015-02-13 22:03:44.514192215 +0100
@@ -360,19 +360,6 @@
 	rrs = &ri->groups[group].rates[rate];
 	tmp_prob_ewma = rrs->prob_ewma;
 
-	/* do not account throughput if sucess prob is below 10% */
-	if (rrs->prob_ewma < RHAPSODY_FRAC(10, 100)) {
-		tp_avg = 0;
-		return tp_avg;
-	}
-
-	/*
-	 * For the throughput calculation, limit the probability value to 90% to
-	 * account for collision related packet error rate fluctuation
-	 */
-	if (rrs->prob_ewma > RHAPSODY_FRAC(90, 100))
-		tmp_prob_ewma = RHAPSODY_FRAC(90, 100);
-
 	if (group != RHAPSODY_CCK_GROUP)
 		nsecs = 1000 * ri->overhead / RHAPSODY_TRUNC(ri->avg_ampdu_len);
 
