Index: compat-wireless-2014-11-04/net/mac80211/rc80211_rhapsody_ht.c
===================================================================
--- compat-wireless-2014-11-04.orig/net/mac80211/rc80211_rhapsody_ht.c	2015-02-13 23:46:20.482843730 +0100
+++ compat-wireless-2014-11-04/net/mac80211/rc80211_rhapsody_ht.c	2015-02-14 00:00:49.392438577 +0100
@@ -186,26 +186,17 @@
  */
 const struct mcs_group rhapsody_mcs_groups[] = {
 	MCS_GROUP(1, 0, BW_20),
-	MCS_GROUP(2, 0, BW_20),
-#if RHAPSODY_MAX_STREAMS >= 3
-	MCS_GROUP(3, 0, BW_20),
-#endif
-
 	MCS_GROUP(1, 1, BW_20),
-	MCS_GROUP(2, 1, BW_20),
-#if RHAPSODY_MAX_STREAMS >= 3
-	MCS_GROUP(3, 1, BW_20),
-#endif
-
 	MCS_GROUP(1, 0, BW_40),
-	MCS_GROUP(2, 0, BW_40),
-#if RHAPSODY_MAX_STREAMS >= 3
-	MCS_GROUP(3, 0, BW_40),
-#endif
-
 	MCS_GROUP(1, 1, BW_40),
+	MCS_GROUP(2, 0, BW_20),
+	MCS_GROUP(2, 1, BW_20),
+	MCS_GROUP(2, 0, BW_40),
 	MCS_GROUP(2, 1, BW_40),
 #if RHAPSODY_MAX_STREAMS >= 3
+	MCS_GROUP(3, 0, BW_20),
+	MCS_GROUP(3, 1, BW_20),
+	MCS_GROUP(3, 0, BW_40),
 	MCS_GROUP(3, 1, BW_40),
 #endif
 
@@ -213,38 +204,23 @@
 
 #ifdef CPTCFG_MAC80211_RC_RHAPSODY_BLUES_VHT
 	VHT_GROUP(1, 0, BW_20),
-	VHT_GROUP(2, 0, BW_20),
-#if RHAPSODY_MAX_STREAMS >= 3
-	VHT_GROUP(3, 0, BW_20),
-#endif
-
 	VHT_GROUP(1, 1, BW_20),
-	VHT_GROUP(2, 1, BW_20),
-#if RHAPSODY_MAX_STREAMS >= 3
-	VHT_GROUP(3, 1, BW_20),
-#endif
-
 	VHT_GROUP(1, 0, BW_40),
-	VHT_GROUP(2, 0, BW_40),
-#if RHAPSODY_MAX_STREAMS >= 3
-	VHT_GROUP(3, 0, BW_40),
-#endif
-
 	VHT_GROUP(1, 1, BW_40),
-	VHT_GROUP(2, 1, BW_40),
-#if RHAPSODY_MAX_STREAMS >= 3
-	VHT_GROUP(3, 1, BW_40),
-#endif
-
 	VHT_GROUP(1, 0, BW_80),
-	VHT_GROUP(2, 0, BW_80),
-#if RHAPSODY_MAX_STREAMS >= 3
-	VHT_GROUP(3, 0, BW_80),
-#endif
-
 	VHT_GROUP(1, 1, BW_80),
+	VHT_GROUP(2, 0, BW_20),
+	VHT_GROUP(2, 1, BW_20),
+	VHT_GROUP(2, 0, BW_40),
+	VHT_GROUP(2, 1, BW_40),
+	VHT_GROUP(2, 0, BW_80),
 	VHT_GROUP(2, 1, BW_80),
 #if RHAPSODY_MAX_STREAMS >= 3
+	VHT_GROUP(3, 0, BW_20),
+	VHT_GROUP(3, 1, BW_20),
+	VHT_GROUP(3, 0, BW_40),
+	VHT_GROUP(3, 1, BW_40),
+	VHT_GROUP(3, 0, BW_80),
 	VHT_GROUP(3, 1, BW_80),
 #endif
 #endif
