From 40536ecddd60a9897f258be183983b02f5296ec2 Mon Sep 17 00:00:00 2001
From: Bastian Bittorf <bittorf@bluebottle.com>
Date: Thu, 8 Oct 2015 17:06:16 +0200
Subject: [PATCH] busybox: add ip neigh menuconfig entry


Signed-off-by: Bastian Bittorf <bittorf@bluebottle.com>
---
 package/utils/busybox/Config-defaults.in |    3 +++
 1 file changed, 3 insertions(+)

diff --git a/package/utils/busybox/Config-defaults.in b/package/utils/busybox/Config-defaults.in
index a61109a..ab28f62 100644
--- a/package/utils/busybox/Config-defaults.in
+++ b/package/utils/busybox/Config-defaults.in
@@ -2119,6 +2119,9 @@ config BUSYBOX_DEFAULT_FEATURE_IP_TUNNEL
 config BUSYBOX_DEFAULT_FEATURE_IP_RULE
 	bool
 	default n
+config BUSYBOX_DEFAULT_FEATURE_IP_NEIGH
+	bool
+	default n
 config BUSYBOX_DEFAULT_FEATURE_IP_SHORT_FORMS
 	bool
 	default n
-- 
1.7.10.4

