From 013042db35f76acf510ffcfc3019a695103a55c5 Mon Sep 17 00:00:00 2001
From: Bastian Bittorf <bittorf@bluebottle.com>
Date: Mon, 18 May 2015 14:38:05 +0200
Subject: [PATCH] enable support local clock and DCF77 - enable full build


Signed-off-by: Bastian Bittorf <bittorf@bluebottle.com>
---
 feeds/packages/net/ntpd/Makefile |    5 +----
 1 file changed, 1 insertion(+), 4 deletions(-)

diff --git a/feeds/packages/net/ntpd/Makefile b/feeds/packages/net/ntpd/Makefile
index 2b64964..5ee6ae6 100644
--- a/feeds/packages/net/ntpd/Makefile
+++ b/feeds/packages/net/ntpd/Makefile
@@ -123,12 +123,9 @@ CONFIGURE_VARS += \
 	ac_cv_header_dns_sd_h=no
 
 CONFIGURE_ARGS += \
-	--disable-all-clocks \
-	--disable-parse-clocks \
-	--without-ntpsnmpd \
-	--without-lineeditlibs \
 	--enable-NMEA \
 	--enable-LOCAL-CLOCK \
+	--enable-RAWDCF \
 	--enable-SHM \
 	--enable-linuxcaps
 
-- 
1.7.10.4

