Index: compat-wireless-YYYY-MM-DD/drivers/net/wireless/ath/regd.c
===================================================================
--- compat-wireless-YYYY-MM-DD.orig/drivers/net/wireless/ath/regd.c	2013-05-11 23:32:07.134034802 +0200
+++ compat-wireless-YYYY-MM-DD/drivers/net/wireless/ath/regd.c	2013-05-11 23:33:32.031500463 +0200
@@ -52,12 +52,11 @@
 #define ATH9K_5GHZ_5725_5850	REG_RULE(5725-10, 5850+10, 40, 0, 30,\
 				NL80211_RRF_PASSIVE_SCAN | NL80211_RRF_NO_IBSS)
 
-#define ATH9K_2GHZ_ALL		ATH9K_2GHZ_CH01_11, \
-				ATH9K_2GHZ_CH12_13, \
-				ATH9K_2GHZ_CH14
+#define ATH9K_2GHZ_ALL		REG_RULE(2400, 2494, 40, 0, 30, 0)
+				
+				
 
-#define ATH9K_5GHZ_ALL		ATH9K_5GHZ_5150_5350, \
-				ATH9K_5GHZ_5470_5850
+#define ATH9K_5GHZ_ALL		REG_RULE(4910, 5895, 40, 0, 30, 0)
 
 /* This one skips what we call "mid band" */
 #define ATH9K_5GHZ_NO_MIDBAND	ATH9K_5GHZ_5150_5350, \
@@ -81,9 +80,8 @@
 static const struct ieee80211_regdomain ath_world_regdom_63_65 = {
 	.alpha2 =  "99",
 	REGD_RULES(
-		ATH9K_2GHZ_CH01_11,
-		ATH9K_2GHZ_CH12_13,
-		ATH9K_5GHZ_NO_MIDBAND,
+		ATH9K_2GHZ_ALL,
+		ATH9K_5GHZ_ALL,
 	)
 };
 
@@ -91,8 +89,8 @@
 static const struct ieee80211_regdomain ath_world_regdom_64 = {
 	.alpha2 =  "99",
 	REGD_RULES(
-		ATH9K_2GHZ_CH01_11,
-		ATH9K_5GHZ_NO_MIDBAND,
+		ATH9K_2GHZ_ALL,
+		ATH9K_5GHZ_ALL,
 	)
 };
 
@@ -100,7 +98,7 @@
 static const struct ieee80211_regdomain ath_world_regdom_66_69 = {
 	.alpha2 =  "99",
 	REGD_RULES(
-		ATH9K_2GHZ_CH01_11,
+		ATH9K_2GHZ_ALL,
 		ATH9K_5GHZ_ALL,
 	)
 };
@@ -109,8 +107,7 @@
 static const struct ieee80211_regdomain ath_world_regdom_67_68_6A_6C = {
 	.alpha2 =  "99",
 	REGD_RULES(
-		ATH9K_2GHZ_CH01_11,
-		ATH9K_2GHZ_CH12_13,
+		ATH9K_2GHZ_ALL,
 		ATH9K_5GHZ_ALL,
 	)
 };
@@ -176,4 +173,4 @@
-	return (center_freq >= 5260 && center_freq <= 5700);
+	return false;
 }
 
 /*
