--- a/net/mac80211/rc80211_minstrel.h
+++ b/net/mac80211/rc80211_minstrel.h
@@ -23,6 +23,12 @@
 /* number of highest throughput rates to consider*/
 #define MAX_THR_RATES 4
 
+/* time in milliseconds the wireless channel is assumed to be stable */
+#define WIFI_COHERENCE_TIME 25
+
+/* minimum number of packets between two successive sample attempts */
+#define MIN_INTERSAMPLE_SPACING 1
+
 /*
  * Perform EWMA (Exponentially Weighted Moving Average) calculation
  */
@@ -91,6 +97,7 @@ struct minstrel_sta_info {
 	struct ieee80211_sta *sta;
 
 	unsigned long last_stats_update;
+	unsigned long last_rate_update;
 	unsigned int sp_ack_dur;
 	unsigned int rate_avg;
 
@@ -98,16 +105,15 @@ struct minstrel_sta_info {
 
 	u8 max_tp_rate[MAX_THR_RATES];
 	u8 max_prob_rate;
+	u8 cur_intersample_spacing;
 	unsigned int total_packets;
 	unsigned int sample_packets;
-	int sample_deferred;
 
 	unsigned int sample_row;
 	unsigned int sample_column;
 
 	int n_rates;
 	struct minstrel_rate *r;
-	bool prev_sample;
 
 	/* sampling table */
 	u8 *sample_table;
@@ -120,14 +126,11 @@ struct minstrel_sta_info {
 
 struct minstrel_priv {
 	struct ieee80211_hw *hw;
-	bool has_mrr;
 	unsigned int cw_min;
 	unsigned int cw_max;
 	unsigned int max_retry;
 	unsigned int segment_size;
-	unsigned int update_interval;
-	unsigned int lookaround_rate;
-	unsigned int lookaround_rate_mrr;
+	bool has_mrr;
 
 	u8 cck_rates[4];
 
--- a/net/mac80211/rc80211_minstrel.c
+++ b/net/mac80211/rc80211_minstrel.c
@@ -88,6 +88,32 @@ int minstrel_get_tp_avg(struct minstrel_
 		return MINSTREL_TRUNC(100000 * (prob_ewma / usecs));
 }
 
+/* Check whether rate sampling is required */
+static inline bool
+minstrel_requires_sampling(struct minstrel_sta_info *mi,
+			   struct minstrel_priv *mp,
+			   bool mrr_capable)
+{
+	int rate_sampling_interval;
+
+	/* Check multi-rate-retry capabilities & adjust sampling interval */
+	if (mrr_capable)
+		rate_sampling_interval = WIFI_COHERENCE_TIME;
+	else
+		rate_sampling_interval = WIFI_COHERENCE_TIME * 2;
+
+	if (mi->cur_intersample_spacing >= MIN_INTERSAMPLE_SPACING &&
+	    (time_after(jiffies, mi->last_rate_update +
+	    (rate_sampling_interval * HZ) / 1000))) {
+		mi->cur_intersample_spacing = 0;
+		mi->last_rate_update = jiffies;
+		return true;
+	} else {
+		mi->cur_intersample_spacing++;
+		return false;
+	}
+}
+
 /* find & sort topmost throughput rates */
 static inline void
 minstrel_sort_best_tp_rates(struct minstrel_sta_info *mi, int i, u8 *tp_list)
@@ -297,11 +323,8 @@ minstrel_tx_status(void *priv, struct ie
 	if ((info->flags & IEEE80211_TX_CTL_RATE_CTRL_PROBE) && (i >= 0))
 		mi->sample_packets++;
 
-	if (mi->sample_deferred > 0)
-		mi->sample_deferred--;
-
 	if (time_after(jiffies, mi->last_stats_update +
-				(mp->update_interval * HZ) / 1000))
+				(WIFI_COHERENCE_TIME * 3 * HZ) / 1000))
 		minstrel_update_stats(mp, mi);
 }
 
@@ -319,7 +342,6 @@ minstrel_get_retry_count(struct minstrel
 	return retry;
 }
 
-
 static int
 minstrel_get_next_sample(struct minstrel_sta_info *mi)
 {
@@ -343,30 +365,25 @@ minstrel_get_rate(void *priv, struct iee
 	struct ieee80211_tx_info *info = IEEE80211_SKB_CB(skb);
 	struct minstrel_sta_info *mi = priv_sta;
 	struct minstrel_priv *mp = priv;
-	struct ieee80211_tx_rate *rate = &info->control.rates[0];
+	struct ieee80211_tx_rate *sampling_setup = NULL;
 	struct minstrel_rate *msr, *mr;
-	unsigned int ndx;
-	bool mrr_capable;
-	bool prev_sample;
-	int delta;
-	int sampling_ratio;
+	unsigned int sampling_ndx = 0;
+	bool indirect_rate_sampling = false;
+	bool mrr_capable = false;
 
 	/* management/no-ack frames do not use rate control */
 	if (rate_control_send_low(sta, priv_sta, txrc))
 		return;
 
-	/* check multi-rate-retry capabilities & adjust lookaround_rate */
-	mrr_capable = mp->has_mrr &&
-		      !txrc->rts &&
-		      !txrc->bss_conf->use_cts_prot;
-	if (mrr_capable)
-		sampling_ratio = mp->lookaround_rate_mrr;
-	else
-		sampling_ratio = mp->lookaround_rate;
-
 	/* increase sum packet counter */
 	mi->total_packets++;
 
+	/* avoid overflowing packet counters */
+	if (mi->total_packets == ~0) {
+		mi->sample_packets = 0;
+		mi->total_packets = 0;
+	}
+
 #ifdef CPTCFG_MAC80211_DEBUGFS
 	if (mp->fixed_rate_idx != -1)
 		return;
@@ -377,73 +394,49 @@ minstrel_get_rate(void *priv, struct iee
 	    (info->control.flags & IEEE80211_TX_CTRL_PORT_CTRL_PROTO))
 		return;
 
-	delta = (mi->total_packets * sampling_ratio / 100) -
-			(mi->sample_packets + mi->sample_deferred / 2);
-
-	/* delta < 0: no sampling required */
-	prev_sample = mi->prev_sample;
-	mi->prev_sample = false;
-	if (delta < 0 || (!mrr_capable && prev_sample))
-		return;
+	mrr_capable = mp->has_mrr && !txrc->rts && !txrc->bss_conf->use_cts_prot;
 
-	if (mi->total_packets >= 10000) {
-		mi->sample_deferred = 0;
-		mi->sample_packets = 0;
-		mi->total_packets = 0;
-	} else if (delta > mi->n_rates * 2) {
-		/* With multi-rate retry, not every planned sample
-		 * attempt actually gets used, due to the way the retry
-		 * chain is set up - [max_tp,sample,prob,lowest] for
-		 * sample_rate < max_tp.
-		 *
-		 * If there's too much sampling backlog and the link
-		 * starts getting worse, minstrel would start bursting
-		 * out lots of sampling frames, which would result
-		 * in a large throughput loss. */
-		mi->sample_packets += (delta - mi->n_rates * 2);
-	}
-
-	/* get next random rate sample */
-	ndx = minstrel_get_next_sample(mi);
-	msr = &mi->r[ndx];
-	mr = &mi->r[mi->max_tp_rate[0]];
-
-	/* Decide if direct ( 1st mrr stage) or indirect (2nd mrr stage)
-	 * rate sampling method should be used.
-	 * Respect such rates that are not sampled for 20 interations.
-	 */
-	if (mrr_capable &&
-	    msr->perfect_tx_time > mr->perfect_tx_time &&
-	    msr->stats.sample_skipped < 20) {
-		/* Only use IEEE80211_TX_CTL_RATE_CTRL_PROBE to mark
-		 * packets that have the sampling rate deferred to the
-		 * second MRR stage. Increase the sample counter only
-		 * if the deferred sample rate was actually used.
-		 * Use the sample_deferred counter to make sure that
-		 * the sampling is not done in large bursts */
-		info->flags |= IEEE80211_TX_CTL_RATE_CTRL_PROBE;
-		rate++;
-		mi->sample_deferred++;
-	} else {
-		if (!msr->sample_limit)
+	/* check if sampling is required */
+	if (unlikely(minstrel_requires_sampling(mi, mp, mrr_capable))) {
+		/* get next random rate sample */
+		sampling_ndx = minstrel_get_next_sample(mi);
+		msr = &mi->r[sampling_ndx];
+
+		/* If we're not using MRR and the sampling rate already
+		* has a probability of >95%, we shouldn't be attempting
+		* to use it, as this only wastes precious airtime */
+		if (!mrr_capable && (mi->r[sampling_ndx].stats.prob_ewma >
+		    MINSTREL_FRAC(95, 100)))
 			return;
 
-		mi->sample_packets++;
-		if (msr->sample_limit > 0)
-			msr->sample_limit--;
-	}
+		/* current max throughput rate */
+		mr = &mi->r[mi->max_tp_rate[0]];
 
-	/* If we're not using MRR and the sampling rate already
-	 * has a probability of >95%, we shouldn't be attempting
-	 * to use it, as this only wastes precious airtime */
-	if (!mrr_capable &&
-	   (mi->r[ndx].stats.prob_ewma > MINSTREL_FRAC(95, 100)))
+		/* Check if direct (1st mrr stage) or indirect (2nd mrr stage)
+		 * rate sampling method should be used. Enforce rate sampling
+		 * for such rates not sampled within last 20 update cycles. */
+		if (mrr_capable &&
+		    msr->perfect_tx_time > mr->perfect_tx_time &&
+		    msr->stats.sample_skipped < 20)
+			indirect_rate_sampling = true;
+
+		/* setup mrr sampling: indirect -> mrr[1], direct -> mrr[0] */
+		if (indirect_rate_sampling) {
+			info->flags |= IEEE80211_TX_CTL_RATE_CTRL_PROBE;
+			sampling_setup = &info->control.rates[1];
+		} else {
+			sampling_setup = &info->control.rates[0];
+			mi->sample_packets++;
+		}
+	} else {
+		/* no sampling required */
 		return;
+	}
 
-	mi->prev_sample = true;
-
-	rate->idx = mi->r[ndx].rix;
-	rate->count = minstrel_get_retry_count(&mi->r[ndx], info);
+	/* assign sampling parameter*/
+	sampling_setup->idx = mi->r[sampling_ndx].rix;
+	sampling_setup->count = minstrel_get_retry_count(&mi->r[sampling_ndx],
+	                                                 info);
 }
 
 
@@ -572,6 +565,7 @@ minstrel_rate_init(void *priv, struct ie
 
 	mi->n_rates = n;
 	mi->last_stats_update = jiffies;
+	mi->last_rate_update = jiffies;
 
 	init_sample_table(mi);
 	minstrel_update_rates(mp, mi);
@@ -606,6 +600,8 @@ minstrel_alloc_sta(void *priv, struct ie
 		goto error1;
 
 	mi->last_stats_update = jiffies;
+	mi->last_rate_update = jiffies;
+
 	return mi;
 
 error1:
@@ -671,12 +667,6 @@ minstrel_alloc(struct ieee80211_hw *hw, 
 	mp->cw_min = 15;
 	mp->cw_max = 1023;
 
-	/* number of packets (in %) to use for sampling other rates
-	 * sample less often for non-mrr packets, because the overhead
-	 * is much higher than with mrr */
-	mp->lookaround_rate = 5;
-	mp->lookaround_rate_mrr = 10;
-
 	/* maximum time that the hw is allowed to stay in one MRR segment */
 	mp->segment_size = 6000;
 
@@ -690,7 +680,6 @@ minstrel_alloc(struct ieee80211_hw *hw, 
 		mp->has_mrr = true;
 
 	mp->hw = hw;
-	mp->update_interval = 100;
 
 #ifdef CPTCFG_MAC80211_DEBUGFS
 	mp->fixed_rate_idx = (u32) -1;
--- a/net/mac80211/rc80211_minstrel_ht.c
+++ b/net/mac80211/rc80211_minstrel_ht.c
@@ -734,7 +734,7 @@ minstrel_ht_tx_status(void *priv, struct
 	}
 
 	if (time_after(jiffies, mi->last_stats_update +
-				(mp->update_interval / 2 * HZ) / 1000)) {
+				(WIFI_COHERENCE_TIME * 2 * HZ) / 1000)) {
 		update = true;
 		minstrel_ht_update_stats(mp, mi);
 	}
