--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -1217,6 +1217,7 @@ enum ieee80211_conf_flags {
  * @IEEE80211_CONF_CHANGE_SMPS: Spatial multiplexing powersave mode changed
  *	Note that this is only valid if channel contexts are not used,
  *	otherwise each channel context has the number of chains listed.
+ * @IEEE80211_CONF_CHANGE_ACK_POWER: TPC power (in dBm) for ACK packets changed
  */
 enum ieee80211_conf_changed {
 	IEEE80211_CONF_CHANGE_SMPS		= BIT(1),
@@ -1227,6 +1228,7 @@ enum ieee80211_conf_changed {
 	IEEE80211_CONF_CHANGE_CHANNEL		= BIT(6),
 	IEEE80211_CONF_CHANGE_RETRY_LIMITS	= BIT(7),
 	IEEE80211_CONF_CHANGE_IDLE		= BIT(8),
+	IEEE80211_CONF_CHANGE_ACK_POWER         = BIT(9),
 };
 
 /**
@@ -1292,6 +1294,7 @@ struct ieee80211_conf {
 	u8 ps_dtim_period;
 
 	u8 long_frame_max_tx_count, short_frame_max_tx_count;
+	s8 ack_power;
 
 	struct cfg80211_chan_def chandef;
 	bool radar_enabled;
