From 157f6a459171a65f9065723cf97e2878bd17cf3b Mon Sep 17 00:00:00 2001
From: Bastian Bittorf <bittorf@bluebottle.com>
Date: Sat, 16 May 2015 16:31:54 +0200
Subject: [PATCH] hotplug: rt2x00 hotplug-script without exit


Signed-off-by: Bastian Bittorf <bittorf@bluebottle.com>
---
 .../etc/hotplug.d/firmware/10-rt2x00-eeprom         |   19 +++++++++++--------
 1 file changed, 11 insertions(+), 8 deletions(-)

diff --git a/target/linux/ramips/base-files/etc/hotplug.d/firmware/10-rt2x00-eeprom b/target/linux/ramips/base-files/etc/hotplug.d/firmware/10-rt2x00-eeprom
index 0cd95a5..aef8f78 100644
--- a/target/linux/ramips/base-files/etc/hotplug.d/firmware/10-rt2x00-eeprom
+++ b/target/linux/ramips/base-files/etc/hotplug.d/firmware/10-rt2x00-eeprom
@@ -2,7 +2,7 @@
 
 rt2x00_eeprom_die() {
 	echo "rt2x00 eeprom: " "$*"
-	exit 1
+	return 1
 }
 
 rt2x00_eeprom_extract() {
@@ -14,19 +14,22 @@ rt2x00_eeprom_extract() {
 	. /lib/functions.sh
 
 	mtd=$(find_mtd_part $part)
-	[ -n "$mtd" ] || \
+	[ -n "$mtd" ] || {
 		rt2x00_eeprom_die "no mtd device found for partition $part"
+		return 1
+	}
 
-	dd if=$mtd of=/lib/firmware/$FIRMWARE bs=1 skip=$offset count=$count 2>/dev/null || \
+	dd if=$mtd of=/lib/firmware/$FIRMWARE bs=1 skip=$offset count=$count 2>/dev/null || {
 		rt2x00_eeprom_die "failed to extract from $mtd"
+		return 1
+	}
 }
 
-FW="/lib/firmware/$FIRMWARE"
-[ -e "$FW" ] && exit 0
+[ -e "/lib/firmware/$FIRMWARE" ] && {
+	. /lib/ramips.sh
 
-. /lib/ramips.sh
-
-board=$(ramips_board_name)
+	board=$(ramips_board_name)
+}
 
 case "$FIRMWARE" in
 "soc_wmac.eeprom")
-- 
1.7.10.4

